CIPs/cip-granda_mento.md

```diff
@@ -0,0 +1,184 @@
+
+The following modifications will be made:
+1. **Modification to [mint()](https://github.com/celo-org/celo-monorepo/blob/master/packages/protocol/contracts/stability/StableToken.sol#L224)**
+   * Allow `msg.sender` to be the stable token's `GrandaMento` in addition to the existing permitted senders, Exchange and `Validators.sol`.
+2. **Modification to [burn()](https://github.com/celo-org/celo-monorepo/blob/master/packages/protocol/contracts/stability/StableToken.sol#L273)**
+   * Allow `msg.sender` to be the stable token's `GrandaMento` in addition to the existing permitted sender, the Exchange.
+
+#### New contracts:
+
+##### `GrandaMento.sol`
+
+A new contract, `GrandaMento.sol`, is created. One deployment of `GrandaMento` exists per stable token. It must have the ability to mint/burn the corresponding stable token and be a spender of `Reserve.sol`.
+
+The contract is owned Governance, is freezable, and has the following configurable parameters:
+
+1. **`address public approver;`** - Set by governance. Intended to be a multisig and has the authority to approve exchanges.
+2. **`uint256 public minStableExchangeAmount;`** - Set by governance. Minimum amount of the stable token an exchange can buy/sell.
+3. **`uint256 public maxStableExchangeAmount;`** - Set by governance. Maximum amount of the stable token an exchange can buy/sell.
+4. **`uint256 public exchangeWaitPeriod;`** - Set by governance. The minimum amount of time that must elapse between a proposed exchange being approved and when the exchange can be executed. Should give sufficient time for Governance to veto an approved exchange.
+5. **`uint256 public spread;`** - Set by governance. The percent fee imposed upon an exchange execution.
+
+The contract has the following functions:
+
+1. **`function proposeExchange(uint256 amount, bool sellCelo) external returns (uint256)`** - Called by an exchange proposer to propose an exchange.
+   * Callable by anyone.
+   * Requires the amount of stable token being bought/sold to be within the range `[minStableExchangeAmount, maxStableExchangeAmount]`.
+   * Deposits the full amount of the asset being sold into the contract.
+   * Records:
+     * The exchange as in the Proposed state.
+     * A struct with the following info is stored in a mapping with an `id` key:
+       * Which asset is being sold.
+       * The amount of the asset being sold.
+   * Returns:
+     * The `id` of the struct in the mapping.
+2. **`function approveExchangeProposal(uint256 id) external`** - Approves a proposed exchange.
+   * Only callable by the approver address.
+   * Marks the proposed exchange with the given `id` as approved.
+   * Records:
+     * The exchange as in the Approved state.
+     * The timestamp at which the approval occurred (ie `block.timestamp`).
+3. **`function cancelExchangeProposal(uint256 id) external`** - Cancels a proposed exchange.
```

Comment:
Cancelling or filling a proposal are the only ways for the proposer to get something back. Should we have some kind of time expiry on these proposals (~1-2 weeks)? I'm imagining a worst case where 2/3 of the multisig holders aren't available, on vacation, etc.

Reply:
If the min amount of multisig signers isn't available, then the proposal hasn't yet been approved by the multisig, which means the proposer is still able to cancel their proposal. Is that sufficient to cover the cases you were thinking of?

If the proposal has been approved, then it will either have to be vetoed by the community or be executed once the delay has passed

Reply:
Ah that makes sense. Resolved.